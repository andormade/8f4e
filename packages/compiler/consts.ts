export const I32_SIGNED_SMALLEST_NUMBER = -2147483648;
export const I16_SIGNED_LARGEST_NUMBER = 32767;
export const I16_SIGNED_SMALLEST_NUMBER = -32768;
