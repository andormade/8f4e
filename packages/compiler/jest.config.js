module.exports = {
	transform: {
		"^.+\\.(t|j)sx?$": ["@swc/jest", {
			jsc: {
				target: 'es2021',
			},
			sourceMaps: true,
		}],
	},
	testMatch: ['/**/*.test.ts'],
};
