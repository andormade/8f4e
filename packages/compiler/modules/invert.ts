import { i32load, i32const, i32store } from '../wasm/instructions';
import { createFunctionBody } from '../wasm/sections';
import { Instruction } from 'wasm-bytecode-utils';
import { ModuleGenerator } from '../types';

const enum Memory {
	ZERO = 0x00,
	INPUT_POINTER = 0x04,
	OUTPUT = 0x08,
}

const invert: ModuleGenerator = function (moduleId, offset) {
	const functionBody = createFunctionBody(
		[],
		[
			...i32const(Memory.OUTPUT + offset),
			...i32const(Memory.INPUT_POINTER + offset),
			...i32load(),
			...i32load(),
			...i32const(-1),
			Instruction.I32_MUL, // TODO: replace MUL with bitwise operation
			...i32store(),
		]
	);

	return {
		moduleId,
		functionBody,
		offset,
		initialMemory: [0, Memory.ZERO + offset, Memory.ZERO + offset, 0],
		memoryAddresses: [
			{ address: Memory.OUTPUT + offset, id: 'out' },
			{ address: Memory.INPUT_POINTER + offset, id: 'in', default: Memory.ZERO + offset, isInputPointer: true },
		],
	};
};

export default invert;
