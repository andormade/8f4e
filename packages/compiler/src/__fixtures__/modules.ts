export default [
	{
		state: {},
		code: `
            module abs
        
            # memory
            memory DEFAULT_VALUE 0
            memory in &DEFAULT_VALUE
            memory out 0
        
            # locals
            local input
        
            # code
            push &out
            push in
            localSet input
            push input
            push 0
            lessThan
            if 
                push 0
                push input
                sub
            else
                push input
            end
            store
        `.split('\n'),
	},
];
