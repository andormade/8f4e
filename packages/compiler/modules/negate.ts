import { i32load, i32const, i32store, ifelse } from '../wasm/instructions';
import { createFunctionBody } from '../wasm/sections';
import { Instruction, Type } from 'wasm-bytecode-utils';
import { ModuleGenerator } from '../types';
import { I16_SIGNED_LARGEST_NUMBER } from '../consts';

const enum Memory {
	ZERO = 0x00,
	INPUT_POINTER = 0x04,
	OUTPUT = 0x08,
}

const negate: ModuleGenerator = function (moduleId, offset) {
	const functionBody = createFunctionBody(
		[],
		[
			...i32const(Memory.OUTPUT + offset),
			...i32const(Memory.INPUT_POINTER + offset),
			...i32load(),
			...i32load(),
			...i32const(0),
			Instruction.I32_GT_S,
			...ifelse(Type.I32, [...i32const(0)], [...i32const(I16_SIGNED_LARGEST_NUMBER)]),
			...i32store(),
		]
	);

	return {
		moduleId,
		functionBody,
		offset,
		initialMemory: [0, Memory.ZERO + offset, Memory.ZERO + offset, 0],
		memoryAddresses: [
			{ address: Memory.OUTPUT + offset, id: 'out' },
			{ address: Memory.INPUT_POINTER + offset, id: 'in', default: Memory.ZERO + offset, isInputPointer: true },
		],
	};
};

export default negate;
