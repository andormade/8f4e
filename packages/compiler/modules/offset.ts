import { i32load, i32const, i32store, ifelse, localSet, localGet } from '../wasm/instructions';
import { createFunctionBody, createLocalDeclaration } from '../wasm/sections';
import { Instruction, Type } from 'wasm-bytecode-utils';
import { ModuleGenerator } from '../types';
import { I16_SIGNED_LARGEST_NUMBER, I16_SIGNED_SMALLEST_NUMBER } from '../consts';

const enum Memory {
	ZERO = 0x00,
	INPUT_POINTER = 0x04,
	OFFSET = 0x08,
	OUTPUT = 12,
}

const enum Locals {
	RESULT,
	__LENGTH,
}

const offset: ModuleGenerator = function (moduleId, offset, initialConfig) {
	const functionBody = createFunctionBody(
		[createLocalDeclaration(Type.I32, Locals.__LENGTH)],
		[
			...i32const(Memory.OUTPUT + offset),
			...[
				...i32const(Memory.INPUT_POINTER + offset),
				...i32load(),
				...i32load(),

				...i32const(Memory.OFFSET + offset),
				...i32load(),

				Instruction.I32_ADD,
				...localSet(Locals.RESULT),

				// Limit
				...localGet(Locals.RESULT),
				...i32const(I16_SIGNED_LARGEST_NUMBER),
				Instruction.I32_GE_S,
				...ifelse(Type.I32, [...i32const(I16_SIGNED_LARGEST_NUMBER)], [...localGet(Locals.RESULT)]),
				...localSet(Locals.RESULT),

				...localGet(Locals.RESULT),
				...i32const(I16_SIGNED_SMALLEST_NUMBER),
				Instruction.I32_LE_S,
				...ifelse(Type.I32, [...i32const(I16_SIGNED_SMALLEST_NUMBER)], [...localGet(Locals.RESULT)]),
			],
			...i32store(),
		]
	);

	return {
		moduleId,
		functionBody,
		offset,
		initialMemory: [0, Memory.ZERO + offset, initialConfig.offset, 0],
		memoryAddresses: [
			{ address: Memory.OUTPUT + offset, id: 'out' },
			{ address: Memory.OFFSET + offset, id: 'offset', default: initialConfig.offset },
			{ address: Memory.INPUT_POINTER + offset, id: 'in', isInputPointer: true, default: Memory.ZERO + offset },
		],
	};
};

export default offset;
