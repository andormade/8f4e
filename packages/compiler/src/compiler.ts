import { createFunctionBody, createLocalDeclaration } from './wasmUtils/sectionHelpers';
import Type from './wasmUtils/type';
import instructions from './instructions';
import { AST, Argument, ArgumentType, CompiledModule, Namespace, MemoryMap } from './types';
import { WORD_LENGTH } from './consts';

export { MemoryTypes, MemoryMap } from './types';

function parseArgument(argument: string): Argument {
	return /^-?[0-9]+$/.test(argument)
		? { value: parseInt(argument, 10), type: ArgumentType.LITERAL }
		: { value: argument, type: ArgumentType.IDENTIFIER };
}

function parseLine(line: string, lineNumber: number): AST[number] {
	const [, instruction, ...args] = line.match(/\s*(\S+)\s*(\S*)\s*(\S*)\s*(\S*)/) || [];

	return {
		lineNumber,
		instruction,
		arguments: args
			.filter(argument => {
				return argument !== '';
			})
			.map(parseArgument),
	};
}

function isComment(line: string): boolean {
	return /\s*#/.test(line);
}

function isValidInstruction(line: string): boolean {
	return /\s*(\S+)\s*(\S*)\s*(\S*)/.test(line);
}

export function compileToAST(module: string[]) {
	return module
		.map((line, index) => [index + 1, line] as [number, string])
		.filter(([, line]) => !isComment(line))
		.filter(([, line]) => isValidInstruction(line))
		.map(([lineNumber, line]) => {
			return parseLine(line, lineNumber);
		});
}

export function compileLine(
	line: AST[number],
	namespace: Namespace,
	startingByteAddress: number
): { byteCode: number[]; namespace: Namespace } {
	if (!instructions[line.instruction]) {
		throw `1001: Unrecognized instruction: '${line.instruction}'`;
	}
	return instructions[line.instruction](line, namespace, startingByteAddress);
}

function getModuleId(ast: AST): string | undefined {
	return (
		ast
			.find(
				line => line.instruction === 'module' && line.arguments[0] && line.arguments[0].type === ArgumentType.IDENTIFIER
			)
			?.arguments[0].value.toString() || undefined
	);
}

export function compile(module: string[], startingByteAddress = 0): CompiledModule {
	const ast = compileToAST(module);
	const moduleId = getModuleId(ast);

	if (!moduleId) {
		throw '1007: Missing module ID';
	}

	let memory: MemoryMap = new Map();
	let locals: string[] = [];
	let consts: Record<string, number> = {};

	const wa = ast
		.reduce((acc, line) => {
			const { byteCode, namespace } = compileLine(line, { locals, memory, consts }, startingByteAddress);
			consts = namespace.consts;
			locals = namespace.locals;
			memory = namespace.memory;
			acc.push(byteCode);
			return acc;
		}, [] as number[][])
		.flat();

	const [, lastMemoryItem = { relativeWordAddress: 0, wordSize: 0 }] = Array.from(memory).pop() || [];

	return {
		moduleId,
		functionBody: createFunctionBody([createLocalDeclaration(Type.I32, locals.length)], wa),
		byteAddress: startingByteAddress,
		wordAddress: startingByteAddress / WORD_LENGTH,
		memoryMap: memory,
		memoryWordSize: lastMemoryItem.relativeWordAddress + lastMemoryItem.wordSize,
		ast,
	};
}
