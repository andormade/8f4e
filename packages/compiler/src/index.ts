import {
	createCodeSection,
	createExportSection,
	createFunctionBody,
	createFunctionExport,
	createFunctionSection,
	createFunctionType,
	createImportSection,
	createMemoryImport,
	createTypeSection,
} from './wasmUtils/sectionHelpers';
import Type from './wasmUtils/type';
import { call, f32store, i32store } from './wasmUtils/instructionHelpers';
import { compileModule, compileToAST } from './compiler';
// import standardLibrary from './standardLibrary/index';
import {
	AST,
	ArgumentLiteral,
	ArgumentType,
	CompileOptions,
	CompiledModule,
	CompiledModuleLookup,
	Module,
	Namespace,
	Namespaces,
} from './types';
import {
	GLOBAL_ALIGNMENT_BOUNDARY,
	I16_SIGNED_LARGEST_NUMBER,
	I16_SIGNED_SMALLEST_NUMBER,
	I32_SIGNED_LARGEST_NUMBER,
} from './consts';
import { ErrorCode, getError } from './errors';
import { sortModules } from './gaphOptimizer';
import { WASM_MEMORY_PAGE_SIZE } from './wasmUtils/consts';

export * from './types';
export { I16_SIGNED_LARGEST_NUMBER, I16_SIGNED_SMALLEST_NUMBER } from './consts';
export { Instruction } from './instructionCompilers';
export { default as instructions } from './instructionCompilers';
export { instructionParser } from './compiler';

const HEADER = [0x00, 0x61, 0x73, 0x6d];
const VERSION = [0x01, 0x00, 0x00, 0x00];

function collectConstants(ast: AST): Namespace['consts'] {
	return Object.fromEntries(
		ast
			.filter(({ instruction }) => instruction === 'const')
			.map(({ arguments: _arguments }) => {
				return [
					_arguments[0].value,
					{
						value: parseFloat(_arguments[1].value.toString()),
						isInteger: (_arguments[1] as ArgumentLiteral).isInteger,
					},
				];
			})
	);
}

function resolveInterModularConnections(compiledModules: CompiledModuleLookup) {
	compiledModules.forEach(({ ast, memoryMap }) => {
		ast.forEach(line => {
			const { instruction, arguments: _arguments } = line;
			if (
				['int*', 'int**', 'float*', 'float**', 'init', 'int'].includes(instruction) &&
				_arguments[0] &&
				_arguments[1] &&
				_arguments[0].type === ArgumentType.IDENTIFIER &&
				_arguments[1].type === ArgumentType.IDENTIFIER &&
				/&(\S+)\.(\S+)/.test(_arguments[1].value)
			) {
				// Remove &
				const [targetModuleId, targetMemoryId] = _arguments[1].value.substring(1).split('.');

				const targetModule = compiledModules.get(targetModuleId);

				if (!targetModule) {
					throw getError(ErrorCode.UNDECLARED_IDENTIFIER, line);
				}

				const targetMemory = targetModule.memoryMap.get(targetMemoryId);

				if (!targetMemory) {
					throw getError(ErrorCode.UNDECLARED_IDENTIFIER, line);
				}

				const memory = memoryMap.get(_arguments[0].value);

				if (memory) {
					memory.default = targetMemory.byteAddress;
				}
			}
		});
	});
}

export function compileModules(modules: AST[], options: CompileOptions): CompiledModule[] {
	let memoryAddress = options.startingMemoryWordAddress;
	const builtInConsts: Namespace['consts'] = {
		I16_SIGNED_LARGEST_NUMBER: { value: I16_SIGNED_LARGEST_NUMBER, isInteger: true },
		I16_SIGNED_SMALLEST_NUMBER: { value: I16_SIGNED_SMALLEST_NUMBER, isInteger: true },
		I32_SIGNED_LARGEST_NUMBER: { value: I32_SIGNED_LARGEST_NUMBER, isInteger: true },
		WORD_SIZE: { value: GLOBAL_ALIGNMENT_BOUNDARY, isInteger: true },
		...options.environmentExtensions.constants,
	};

	const namespaces: Namespaces = new Map(
		modules.map(ast => {
			const moduleName = getModuleName(ast);
			return [moduleName, { consts: collectConstants(ast) }];
		})
	);

	return modules.map((ast, index) => {
		const module = compileModule(
			ast,
			builtInConsts,
			namespaces,
			memoryAddress * GLOBAL_ALIGNMENT_BOUNDARY,
			options.maxMemorySize,
			index
		);
		memoryAddress += module.wordAlignedSize;

		if (options.maxMemorySize * WASM_MEMORY_PAGE_SIZE <= memoryAddress) {
			throw 'Memory limit exceeded';
		}

		return module;
	});
}

export function getModuleName(ast: AST) {
	const moduleInstruction = ast.find(line => {
		return line.instruction === 'module';
	});

	if (!moduleInstruction) {
		throw 'Missing module instruction';
	}

	const argument = moduleInstruction.arguments[0];

	if (argument.type !== ArgumentType.IDENTIFIER) {
		throw 'Module instruction argument type invalid';
	}

	return argument.value;
}

export function generateMemoryInitiatorFunctions(compiledModules: CompiledModule[]) {
	return compiledModules.map(module => {
		let pointer = module.byteAddress;
		const instructions: number[] = [];

		Array.from(module.memoryMap.values()).forEach(memory => {
			if (memory.numberOfElements > 1 && memory.default instanceof Map) {
				memory.default.forEach((value, relativeWordAddress) => {
					instructions.push(
						...(memory.isInteger
							? i32store(pointer + (relativeWordAddress + 1) * GLOBAL_ALIGNMENT_BOUNDARY, value)
							: f32store(pointer + (relativeWordAddress + 1) * GLOBAL_ALIGNMENT_BOUNDARY, value))
					);
				});
				pointer += memory.wordAlignedSize * GLOBAL_ALIGNMENT_BOUNDARY;
			} else if (memory.numberOfElements === 1 && memory.default !== 0) {
				instructions.push(
					...(memory.isInteger
						? i32store(pointer, memory.default as number)
						: f32store(pointer, memory.default as number))
				);
				pointer += GLOBAL_ALIGNMENT_BOUNDARY;
			} else if (memory.numberOfElements === 1 && memory.default === 0) {
				pointer += GLOBAL_ALIGNMENT_BOUNDARY;
			}
		});

		return createFunctionBody([], instructions);
	});
}

export default function compile(
	modules: Module[],
	options: CompileOptions
): {
	codeBuffer: Uint8Array;
	compiledModules: CompiledModuleLookup;
	allocatedMemorySize: number;
} {
	const astModules = modules.map(({ code }) => compileToAST(code, options));
	// const astFunctions = standardLibrary.map(code => compileToAST(code.split('\n'), options));
	const sortedModules = sortModules(astModules);
	const compiledModules = compileModules(sortedModules, { ...options, startingMemoryWordAddress: 1 });
	const compiledModulesMap = new Map(compiledModules.map(({ id, ...rest }) => [id, { id, ...rest }]));
	resolveInterModularConnections(compiledModulesMap);
	const loopFunctionBodies = compiledModules.map(({ loopFunctionBody }) => loopFunctionBody);
	// const initFunctionBodies = compiledModules.map(({ initFunctionBody }) => initFunctionBody);
	const functionSignatures = compiledModules.map(() => 0x00);
	const cycleFunction = compiledModules.map((module, index) => call(index + 3)).flat();
	const memoryInitiatorFunction = compiledModules
		.map((module, index) => call(index + compiledModules.length + 3))
		.flat();
	const memoryInitiatorFunctions = generateMemoryInitiatorFunctions(compiledModules);

	return {
		codeBuffer: Uint8Array.from([
			...HEADER,
			...VERSION,
			...createTypeSection([
				createFunctionType([], []),
				createFunctionType([Type.I32], [Type.I32]),
				createFunctionType([Type.I32, Type.I32], [Type.I32]),
			]),
			...createImportSection([
				createMemoryImport('js', 'memory', options.initialMemorySize, options.maxMemorySize, true),
			]),
			...createFunctionSection([0x00, 0x00, 0x00, ...functionSignatures, ...functionSignatures]),
			...createExportSection([
				createFunctionExport('init', 0x00),
				createFunctionExport('cycle', 0x01),
				createFunctionExport('buffer', 0x02),
			]),
			...createCodeSection([
				createFunctionBody([], memoryInitiatorFunction),
				createFunctionBody([], cycleFunction),
				createFunctionBody([], new Array(128).fill(call(1)).flat()),
				...loopFunctionBodies,
				...memoryInitiatorFunctions,
			]),
		]),
		compiledModules: compiledModulesMap,
		allocatedMemorySize:
			compiledModules[compiledModules.length - 1].byteAddress +
			compiledModules[compiledModules.length - 1].wordAlignedSize * GLOBAL_ALIGNMENT_BOUNDARY,
	};
}
