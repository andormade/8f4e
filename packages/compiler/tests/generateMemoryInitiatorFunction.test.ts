import modules from './__fixtures__/modules';

import { compileToAST } from '../src/compiler';
import { compileModules, generateMemoryInitiatorFunctions } from '../src';

describe('compiler', () => {
	test('generateMemoryInitiatorFunction', () => {
		const astModules = modules.map(({ code }) => compileToAST(code));
		const compiledModules = compileModules(astModules, {
			startingMemoryWordAddress: 0,
			environmentExtensions: { constants: {}, ignoredKeywords: [] },
			maxMemorySize: 1,
			initialMemorySize: 1,
		});
		expect(generateMemoryInitiatorFunctions(compiledModules)).toMatchSnapshot();
	});
});
