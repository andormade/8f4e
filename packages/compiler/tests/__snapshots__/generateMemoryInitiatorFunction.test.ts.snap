// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`compiler generateMemoryInitiatorFunction 1`] = `
Array [
  Array [
    9,
    0,
    65,
    12,
    65,
    1,
    54,
    2,
    0,
    11,
  ],
]
`;
